### DISCUSSION TITLE
v2.x.x

### DISCUSSION THREAD

**Discussion by addiejanssen** [addiejanssen/domoticz-solaredge-modbustcp-plugin#48] [Releases]
This is the second **pre-release** of the 2.x.x version of the plugin which will support reading `meter` and `battery` data next to the `inverter` data.

This is a major rewrite of the plugin and is only tested by a very small group of users so far.
And although it was tested by them, we may not yet have found all issues with it.

We would love to receive feedback on this version in the Discussions before we are going to release it to the public!

Enjoy!

Addie

## We need your help!
Do you have a meter or battery connected to your inverter?
Then please set the log level to verbose and share the output of your inverter with us!
This allows us to make sure that we have valid/real life data that the plugin should be able to work with.

### Added

Tested with the latest version (0.7.2) of the solaredge_modbus library.

### Changed

During testing it was found that certain inverters return battery information while there were no batteries attached to the inverter.
First investigations shows that this might be an issue caused by the solaredge_modbus library.
We will contact the developer of that library to further investigate.
**For now, battery support has been turned off.**

### Fixed

- In certain situations, the plugin could run into uninitialized variables.
- The timeout value used for communication with the inverter has been increased (from 5 to 15 seconds).

<hr /><em>This discussion was created from the release <a href='https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/releases/tag/2.0.2'>2.0.2</a>.</em>

**Comment by PeterVermaas** [comment]
Hi Addie, 
Thank you for the great work done till now! I couldn't find the 0.7.2 version of solaredge_modbus library but it looks that 2.0.2 also works with 0.7.0? I changed the requirements.txt after which i succeeded to install. I saw in plugin.py and batteries.py that battery is still not worked out.
I intend to get data from the Tesla battery connected to a SE3500. Under chapter Ideas i commented and added a textfile extracted from a slightly adapted example.py containing the Domoticz log data. I hope you want to take a look to the issue. It would help me very much in regulating the energyflows via Domoticz. 
[tesladata.txt](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/12040661/tesladata.txt)

**Reply by addiejanssen** [reply]
Hi Peter,

Looking at the data you supplied, it shows information for 1 battery that sort of makes sense, but the other doesn't. This is one of the issues I have with the solaredge_modbus library: it shows "ghost" information for certain devices.

I was working on battery support and there is a newer version for that purpose: [domoticz-solaredge-modbustcp-plugin.2.0.3.zip](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/12042701/domoticz-solaredge-modbustcp-plugin.2.0.3.zip).
If you want, give it a go and let us know what happens.

For debugging purposes, it would help to set the loglevel to "Verbose". That will generate just enough information to see what is going on.
I am pretty sure that the information reported by the plugin is not yet correct. I was struggling with the "direction" of energy flows. Like: is the energy going to or coming from the battery. To get those kind of information correctly processed, I need feedback from somebody that actually has a battery attached.

The configuration of the plugin provides an option to select which types of hardware it should try to find:
- Inverter
- Inverter+Meters
- Inverter+Batteries
- Inverter+Meters+Batteries

![domoticz-solaredge-modbustcp-plugin 2 0 3](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/assets/55428983/dae7090c-3ab7-48c0-ae17-140fb2ff32e2)

Enjoy,

Addie

**Comment by PeterVermaas** [comment]
Hi Addie,

Thank you for your very fast response, super! I surely give it a try. 

Somehow I’m not able to retrieve the 0.7.2 version so had to fall back on 0.7.0.

 

***@***.***:~/domoticz/plugins/domoticz-solaredge-modbustcp-plugin.2.0.2 $ sudo service domoticz stop

***@***.***:~/domoticz/plugins/domoticz-solaredge-modbustcp-plugin.2.0.2 $ cd ..

***@***.***:~/domoticz/plugins $ cd domoticz-solaredge-modbustcp-plugin.2.0.3

***@***.***:~/domoticz/plugins/domoticz-solaredge-modbustcp-plugin.2.0.3 $ sudo pip3 install -r requirements.txt

Collecting solaredge_modbus==0.7.2 (from -r requirements.txt (line 1))

  Could not find a version that satisfies the requirement solaredge_modbus==0.7.2 (from -r requirements.txt (line 1)) (from versions: 0.1.1, 0.2.0, 0.3.0, 0.4.0, 0.5.0, 0.6.0, 0.6.1, 0.6.2, 0.6.3, 0.6.4, 0.6.5, 0.6.6, 0.7.0)

No matching distribution found for solaredge_modbus==0.7.2 (from -r requirements.txt (line 1))

 

I installed again and removed the older plugin-folders. Verbose logging and Inverter+Batteries are activated in the menu, see below.



 

 

 

In batteries.py, table OTHER_BATTERY, the modbusnames from the last 16 (so from “rated_energy” … “soe”) I saw them when calling via example.py:

 

Battery Registers:

{'c_manufacturer': 'TESLA\x14\tT15L0001124', 'c_model': '1067000-00-B', 'c_version': '2.30.5', 'c_serialnumber': 'T15L0001124', 'c_deviceaddress': 24, 'c_sunspec_did': 0, 'rated_energy': 6400.0, 'maximum_charge_continuous_power': 3300.0, 'maximum_discharge_continuous_power': 3300.0, 'maximum_charge_peak_power': 0.0, 'maximum_discharge_peak_power': 3300.0, 'average_temperature': 30.0, 'maximum_temperature': 0.0, 'instantaneous_voltage': 427.29998779296875, 'instantaneous_current': 0.0, 'instantaneous_power': 0.0, 'lifetime_export_energy_counter': 9992572, 'lifetime_import_energy_counter': 10288016, 'maximum_energy': 6400.0, 'available_energy': 5625.0, 'soh': 87.890625, 'soe': 100.0, 'status': 6, 'status_internal': 3, 'event_log': 0, 'event_log_internal': 0, 'battery_stored_energy': 5625.0}

 

Battery Registers:

{'c_manufacturer': '\x02', 'c_model': 'False', 'c_version': 'False', 'c_serialnumber': 'False', 'c_deviceaddress': 16, 'c_sunspec_did': 0, 'rated_energy': -3.4028234663852886e+38, 'maximum_charge_continuous_power': -3.4028234663852886e+38, 'maximum_discharge_continuous_power': -3.4028234663852886e+38, 'maximum_charge_peak_power': -3.4028234663852886e+38, 'maximum_discharge_peak_power': -3.4028234663852886e+38, 'average_temperature': -3.4028234663852886e+38, 'maximum_temperature': 0.0, 'instantaneous_voltage': -3.4028234663852886e+38, 'instantaneous_current': -3.4028234663852886e+38, 'instantaneous_power': 0.0, 'lifetime_export_energy_counter': 0, 'lifetime_import_energy_counter': 0, 'maximum_energy': -3.4028234663852886e+38, 'available_energy': -3.4028234663852886e+38, 'soh': -3.4028234663852886e+38, 'soe': -3.4028234663852886e+38, 'status': 7, 'status_internal': 0, 'event_log': 0, 'event_log_internal': 0, 'battery_stored_energy': 1.1579207543382391e+75}

 

 

- However I didn’t see STATUS, STATUS_INTERNAL, EVENT_LOG and EVEN_LOG_INTERNAL. Maybe a point of investigation?

- I also noticed in example.py that the c_serialnumber from the “fake” battery contains ‘False”. Is this somewhat you can use in a workaround?

In v2.0.2 row 493 you had something like    #                                to_log = battery_values

#                                if "c_serialnumber" in to_log:

#                                    to_log.pop("c_serialnumber")

#                                DomoLog(LogLevels.VERBOSE, "device: {} values: {}".format("Battery", json.dumps(to_log, indent=4, sort_keys=False)))  #was battery zonder ""

 

 

I also include the Verbose++ logfile from Domoticz. Note that there is no sign of battery info available. You will also find another SolarEdgeModBusTCP: (SolarEdgeModBusTCP), this is another device converting solar energy on the shed. This one still is okay, even after updating to v2.0.3.

 

I hope you find this information useful. I can provide more logs or do some trials. The SolarEdge SE3500 also has -next to the Tesla Powerwall 1- also 2 energymeters attached which already are present in the Domoticz.

 

At last I tried to remove the hardware called Tesla and add a new hardware again but then also no recognition of de battery took place.

**Reply by addiejanssen** [reply]
Hi Peter,

I am not sure if I fully understand your response, so let me check a few things.

You suggest in your response that you included a log, but there is nothing here in GitHub.
I am of course interested in that log and screenshots of what Domoticz shows about the battery.
But looking at your message, I think the Tesla battery is not recognized by the plugin.
Is that correct?

You mention: "However I didn’t see STATUS, STATUS_INTERNAL, EVENT_LOG and EVEN_LOG_INTERNAL. Maybe a point of investigation?".
They are clearly visible in the 2 Battery Registers examples above that line.
Does that mean, that the example.py reports these values but the Domoticz plugin doesn't?

Kind regards,

Addie

**Reply by PeterVermaas** [reply]
Hello Addie,
I tried to get 2.0.3 running but somehow it was not able to retrieve any data from battery and meters. I looked into the code from plugin.py v2.0.3 around row 165-168 and found the bug: 
Parameters["Mode6"] is an int so need to be converted to int before compare:

`if int(Parameters["Mode6"]) == 1 or int(Parameters["Mode6"]) == 3:
            self.scan_for_meters = True
if int(Parameters["Mode6"]) == 2 or int(Parameters["Mode6"]) == 3:
            self.scan_for_batteries = True`

After repairing this the battery and meters popped up in Domoticz.
Another small thing: batteries.py row 57 and 58: FORMAT, SOH and SOE changed to "{:.2f}"  . Now it shows percentage of SOH and SOE in a correct way.

About the Status, Status_Internal etc: I overlooked those items, they are available so nothing to concern about this. I'm pretty new in Github so here another try to upload the files...
[domoticz.log](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/12129602/domoticz.log)
![TeslaInDomoticz](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/assets/119419033/9f11f838-adbb-4b4f-bc04-8444a4182ab1)
If you need anymore help, please let me know!

**Reply by addiejanssen** [reply]
Nice to see the results.

You got me puzzled with the mode6 parameter, because I knew that I already fixed that as shown in the [repository](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/blob/91be72512ebd76df56dc7e6639331ea8f6c931b1/plugin.py#L164C11-L164C11):
```python
        components = int(Parameters["Mode6"]) if Parameters["Mode6"] else 0
        if components == 1:
            self.scan_for_meters = True
        elif components == 2:
            self.scan_for_batteries = True
        elif components == 3:
            self.scan_for_meters = True
            self.scan_for_batteries = True
        else:
            self.scan_for_meters = False
            self.scan_for_batteries = False
```
It took me some time to find out that the zip file that I shared with you didn't have that fix yet.
So good spot!

I have updated the batteries table to match your changes.

Based on this, I have bumped the version up to 2.0.4 and have attached that [version here](https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/files/12136665/domoticz-solaredge-modbustcp-plugin.2.0.4.zip).

What would be interesting to know if the graphs are now correct.
With a battery we can import and export energy, but I don't know if the graphs are now matching the "direction" of the energy flow.
So you could help me by looking really careful at all values and graphs and provide feedback on them.

I think there is an error in the solaredge_modbus library with regards to the status fields.
So also there you can help out: is the status as reported by the plugin correct?

Thanks for the feedback!

**Reply by hjzwiers** [reply]
I wonder how to download the file directly into the plugins directory. I have a solar edge with battery and metering.

I followed the directions for the download: git clone https://github.com/addiejansen/domotica-solaredge-modbustcp-plugin.git

cd

sudo pip3 install -r requirements.txt

and restart of domoticz

But now I am wondering how to download the zip file?

I also ran

sudo pip3 uninstall pymodbus
sudo pip3 install pymodbus==2.4.0

and got a mor complete readout. Please help.

**Reply by addiejanssen** [reply]
Hey @hjzwiers,

The "git clone" instructions are for the 1.x.x version of the plugin which does not support meters and batteries.
The installation instructions for the 2.x.x can be found here: https://github.com/addiejanssen/domoticz-solaredge-modbustcp-plugin/tree/meters

Hope this helps.

**Comment by hjzwiers** [comment]
Hoi Addie,

 

Ik heb alles opniew gedaan en krijg de volgende error is mij logging:

 

2023-09-02 13:05:04.963 Error: Solaredge ModbusTCP: (SolarEdge_ModbusTCP) failed to load 'plugin.py', Python Path used was '/home/hjz/domoticz/plugins/domoticz-solaredge-modbustcp-plugin/:/usr/lib/python39.zip:/usr/lib/python3.9:/usr/lib/python3.9/lib-dynload:/usr/local/lib/python3.9/dist-packages:/usr/lib/python3/dist-packages:/usr/lib/python3.9/dist-packages'.

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: Traceback (most recent call last):

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: File "/home/hjz/domoticz/plugins/domoticz-solaredge-modbustcp-plugin/plugin.py", line 68, in <module>

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: import solaredge_modbus

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: File "/usr/local/lib/python3.9/dist-packages/solaredge_modbus/__init__.py", line 157, in <module>

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: class SolarEdge:

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: File "/usr/local/lib/python3.9/dist-packages/solaredge_modbus/__init__.py", line 163, in SolarEdge

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: wordorder = Endian.Big

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: File "/usr/lib/python3.9/enum.py", line 405, in __getattr__

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: raise AttributeError(name) from None

2023-09-02 13:05:04.969 Error: Solaredge ModbusTCP: AttributeError: Big

 

Wat gaat er nog mis?

 

Mvg

HJZ
